contracts/0.4.24/nos/NodeOperatorsRegistry.sol

```diff
@@ -185,181 +197,203 @@ contract NodeOperatorsRegistry is INodeOperatorsRegistry, IsContract, AragonApp
+        external
         authP(SET_NODE_OPERATOR_LIMIT_ROLE, arr(_id, uint256(_stakingLimit)))
         operatorExists(_id)
     {
         require(operators[_id].stakingLimit != _stakingLimit, "NODE_OPERATOR_STAKING_LIMIT_IS_THE_SAME");
         _increaseKeysOpIndex();
-        operators[_id].stakingLimit = _stakingLimit;
-        emit NodeOperatorStakingLimitSet(_id, _stakingLimit);
+        _setNodeOperatorStakingLimit(_id, _stakingLimit);
+        _updateTotalAvailableKeysCount();
     }
 
     /**
-      * @notice Report `_stoppedIncrement` more stopped validators of the node operator #`_id`
-      */
-    function reportStoppedValidators(uint256 _id, uint64 _stoppedIncrement) external
+     * @notice Report `_stoppedIncrement` more stopped validators of the node operator #`_id`
+     */
+    function reportStoppedValidators(uint256 _id, uint64 _stoppedIncrement)
+        external
         authP(REPORT_STOPPED_VALIDATORS_ROLE, arr(_id, uint256(_stoppedIncrement)))
         operatorExists(_id)
     {
         require(0 != _stoppedIncrement, "EMPTY_VALUE");
         operators[_id].stoppedValidators = operators[_id].stoppedValidators.add(_stoppedIncrement);
         require(operators[_id].stoppedValidators <= operators[_id].usedSigningKeys, "STOPPED_MORE_THAN_LAUNCHED");
 
+        uint64 currentStakingLimit = operators[_id].stakingLimit;
+        if (currentStakingLimit > 0) {
+            uint64 stakingLimitDecrease = currentStakingLimit > _stoppedIncrement ? _stoppedIncrement : currentStakingLimit;
+            _setNodeOperatorStakingLimit(_id, operators[_id].stakingLimit.sub(stakingLimitDecrease));
+        }
+
+        _setTotalActiveKeys(keysUsageStats.totalActiveKeys.sub(_stoppedIncrement));
+
         emit NodeOperatorTotalStoppedValidatorsReported(_id, operators[_id].stoppedValidators);
     }
 
     /**
-      * @notice Remove unused signing keys
-      * @dev Function is used by the Lido contract
-      */
-    function trimUnusedKeys() external onlyLido {
+     * @notice Remove unused signing keys
+     * @dev Function is used by the StakingRouter where WC changed
+     */
+    function trimUnusedKeys() external auth(TRIM_UNUSED_KEYS_ROLE) {
+        uint256 trimmedKeys = 0;
         uint256 length = getNodeOperatorsCount();
         for (uint256 operatorId = 0; operatorId < length; ++operatorId) {
-            uint64 totalSigningKeys = operators[operatorId].totalSigningKeys;
-            uint64 usedSigningKeys = operators[operatorId].usedSigningKeys;
-            if (totalSigningKeys != usedSigningKeys) { // write only if update is needed
-                operators[operatorId].totalSigningKeys = usedSigningKeys;  // discard unused keys
-                emit NodeOperatorTotalKeysTrimmed(operatorId, totalSigningKeys - usedSigningKeys);
-            }
+            trimmedKeys += _trimUnusedNodeOperatorKeys(operatorId);
+        }
+        _updateTotalAvailableKeysCount();
```

Comment:
maybe should call this conditionally as an expression below?

Reply:
fixed